qutip/qip/circuit.py

```diff
@@ -1147,3 +1147,181 @@ def qasm(self):
             code += "\n"
 
         return code
+
+    @classmethod
+    def read_qasm(cls, file_path):
+        """
+        Import a QSAM 1.0 file and return an equivalent QuTiP quantum
+        circuit object of the class QubitCircuit.
+
+        Parameters
+        ----------
+        file_path : string
+            Path to the file.
+
+        Returns
+        -------
+        qc : QubitCircuit
+            Return equivalent qubit circuit object.
+        """
+        # Defining the gates
+        qubit_registary = ["qubits"]
+        static_gate = {'h': 'SNOT', 'x': rx(np.pi), 'y': ry(np.pi),
+                       'z': rz(np.pi), 'x90': rx(np.pi/2), 'y90': ry(np.pi/2),
+                       'z90': rz(np.pi/2), 'mx90': rx(-np.pi/2),
+                       'my90': ry(-np.pi/2), 'mz90': rz(-np.pi/2),
+                       's': phasegate(np.pi/2), 'sdag': phasegate(-np.pi/2),
+                       't': phasegate(np.pi/4), 'tdag': phasegate(-np.pi/4)}
+        param_gate = {'rx': 'RX', 'ry': 'RY', 'rz': 'RZ'}
+        c_gate = {'cnot': 'CNOT', 'cz': 'CSIGN'}
+        crk = {'crk': 'CPHASE'}
+        cr = {'cr': 'CPHASE'}
+        toffoli_gate = {'toffoli': 'TOFFOLI'}
+        swap_gate = {'swap': 'SWAP'}
+
+        gates_keys = []
+        gates_keys.extend(static_gate.keys())
+        gates_keys.extend(param_gate.keys())
+        gates_keys.extend(c_gate.keys())
+        gates_keys.extend(crk.keys())
+        gates_keys.extend(cr.keys())
+        gates_keys.extend(toffoli_gate.keys())
+        gates_keys.extend(swap_gate.keys())
+
+        f = open(file_path, "r")
+        file_data = f.readlines()
+        f.close()
+
+        lines = []
+        for line in file_data:
+            # Striping the code into tokens and keywords
+            line = line.rstrip('\n')
+            line = line.replace('{', '').replace('}', '')
+            line = [x.replace(']', ' ') for x in line.split("|")]
+            line = [x.split() for x in line]
+
+            for command in line:
+                if command:
+                    command = [x.split('[')[-1] for x in command]
+                    command = [x.replace(',', ' ') for x in command]
+                    command = [x.split() for x in command]
+                    command = [x for x in command if x]
+
+                    lines.append(command)
+
+                    # Checking for the size of the qubit register.
+                    if command[0][0] in qubit_registary:
+                        N = int(command[1][0])
+                        qc = QubitCircuit(N)
```

Comment:
If this is a class method, then calls to `QubitCircuit` should be via the parameter `cls` so that subclassed versions behave correctly.  Alternatively, you can remove this from the class and make it a regular function, since Python allows this, and subclassed versions may modify the `__init__` method.

Reply:
Thanks.